cmd/experimental/kjobctl/pkg/builder/templates/slurm_init_entrypoint_script.sh.tmpl

```diff
@@ -35,29 +36,29 @@ SBATCH_JOB_NAME={{.SbatchJobName}}
 SBATCH_PARTITION={{.SbatchPartition}}
 EOF
 {{if .FirstNodeIP}}
-  apk --update add curl jq
-
-  token=$(cat /var/run/secrets/kubernetes.io/serviceaccount/token)
-  ca_cert=/var/run/secrets/kubernetes.io/serviceaccount/ca.crt
-  timeout={{.FirstNodeIPTimeoutSeconds}}
-  start_time=$(date +%s)
-
-  while true; do
-    ip=$(curl -k --cacert $ca_cert -H "Authorization: Bearer $token" "https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT/api/v1/namespaces/{{.Namespace}}/pods?labelSelector=batch.kubernetes.io/job-name={{.JobName}},batch.kubernetes.io/job-completion-index=0" | jq -r '.items[0].status.podIP')
-    if [[ -n "$ip" ]]; then
-      SLURM_JOB_FIRST_NODE_IP=$ip
-      break
-    else
-      current_time=$(date +%s)
-      elapsed_time=$((current_time - start_time))
-      if (( elapsed_time >= timeout )); then
-        echo "timeout reached, IP ip for the first node ($SLURM_JOB_FIRST_NODE) not found."
+  if [[ "$JOB_COMPLETION_INDEX" -eq 0 ]]; then
+    SLURM_JOB_FIRST_NODE_IP=${POD_IP}
+  else
+    apk add --update bind-tools
```

Comment:
What is this needed for?

Reply:
To install nslookup

Reply:
It would be better to have an image which pre-intalls it. Could you check if there are any images with the necessary tools in registry.k8s.io, see: https://explore.ggcr.dev/?repo=registry.k8s.io

Reply:
We looked with @mbobrovskyi at the available images, and it seems that if we want to use nslookup then we need to install `bind-tools` at runtime, as here. 

An alternative is to use ping which appears to work OOTB, but maybe it could trigger traffic to external DNS servers (unlikely since the entries are locally, but I'm not sure).

We could have a dedicated image for kjobctl, but it will require more time and maintenance cost.

WDYT @mwielgus @mwysokin ?

Reply:
> We could have a dedicated image for kjobctl, but it will require more time and maintenance cost.

I would prefer to have a separate image. In that case, we will be independent in the future and can install additional tools if needed. For example, in the previous solution, we needed to install `curl` and `jq`. With a separate image, it would be much easier to install them instead of searching suitable one from existing list and trying to combine.

Reply:
I don't think we should be installing anything if nslookup is part of `busybox`

```sh
docker run -it busybox nslookup
Unable to find image 'busybox:latest' locally
latest: Pulling from library/busybox
a46fbb00284b: Pull complete
Digest: sha256:768e5c6f5cb6db0794eec98dc7a967f40631746c32232b78a3105fb946f3ab83
Status: Downloaded newer image for busybox:latest
BusyBox v1.37.0 (2024-09-26 21:31:42 UTC) multi-call binary.

Usage: nslookup [-type=QUERY_TYPE] [-debug] HOST [DNS_SERVER]

Query DNS about HOST

QUERY_TYPE: soa,ns,a,aaaa,cname,mx,txt,ptr,srv,any
```

Reply:
On busybox we don't have `bash` that is required for us.

Reply:
We also have it current image. But we need to install extra library `bind-tools` to allow DNS resolving on nslookup. Otherwise we have an error:

```
nslookup: can't resolve 'slurm-profile-slurm-cvvlf-0.slurm-profile-slurm-cvvlf': Name does not resolve
```

Reply:
What's the current image being used? 

Environments which have even the lowest security awareness consideration (really any enterprise environment) will be run with:

```yaml
...
    securityContext:
      runAsNonRoot: true
      allowPrivilegeEscalation: false
```

I don't think we should be modifying an image during runtime.

Reply:
> What's the current image being used?

registry.k8s.io/alpine-with-bash:1.0

Reply:
Does the nslookup at busybox still require installing bind-tools? 

If not, then maybe we can find a workaround for the bash requirement [here](https://github.com/kubernetes-sigs/kueue/pull/3072/files#diff-72c2ec4536975c1634d02fff41ef755788d6769d9ffb68b5d6826811b5c28338L14), and use busybox.

Reply:
> Does the nslookup at busybox still require installing bind-tools?

Yes in `busybox` it's working fine.

Reply:
> If not, then maybe we can find a workaround for the bash requirement [here](https://github.com/kubernetes-sigs/kueue/pull/3072/files#diff-72c2ec4536975c1634d02fff41ef755788d6769d9ffb68b5d6826811b5c28338L14), and use busybox.

I would prefer to use bash and create custom image.

Reply:
this option remains on the table, but maybe the workaround to use busybox isn't that hard?

Reply:
IIUC the code 

```
declare -A array_indexes=({{.ArrayIndexes}}) 	# Requires bash v4+
	container_indexes=${array_indexes[${JOB_COMPLETION_INDEX}]}
	container_indexes=(${container_indexes//,/ })
	...
	SLURM_ARRAY_TASK_ID=${container_indexes[$i]}
```
to get the value at the JOB_COMPLETION_INDEX position of comma-separated string. Maybe we could use awk instead for this, which is present at busybox.

Reply:
@mbobrovskyi What are the exact requirements for the image?

- bash
- nslookup/dig/host

Anything else?

Reply:
I think bash may not be needed, as posted above. I think we can extract I-value from csv using awk or 'cut -d,'

Reply:
Created follow up to use busybox https://github.com/kubernetes-sigs/kueue/pull/3238/.

Reply:
That's OK, but maybe it's simpler to find the right image other than modify the script? Whatever takes less time I guess. This is the outcome of my little research:

```sh
$ docker run -it registry.k8s.io/ubuntu-slim:0.14 grep --version
grep (GNU grep) 2.25
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Mike Haertel and others, see <http://git.sv.gnu.org/cgit/grep.git/tree/AUTHORS>.

$ docker run -it registry.k8s.io/ubuntu-slim:0.14 awk --help
awk: not an option: --help

$ docker run -it registry.k8s.io/ubuntu-slim:0.14 bash --version
GNU bash, version 4.3.48(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

$ docker run -it registry.k8s.io/ubuntu-slim:0.14 getent --version
getent (Ubuntu GLIBC 2.23-0ubuntu9) 2.23
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Thorsten Kukuk.
```

0.14 seems to be the newest version, but it's 14.04 from 2017. Having said that busybox also has the latest version from 2017 🤔 I wonder if that could be an issue for our users.

BTW getent can be used in the following way:

```sh
$ docker run -it registry.k8s.io/ubuntu-slim:0.14 getent ahosts google.com
108.177.127.139 STREAM google.com
108.177.127.139 DGRAM
108.177.127.139 RAW
108.177.127.100 STREAM
108.177.127.100 DGRAM
108.177.127.100 RAW
108.177.127.113 STREAM
108.177.127.113 DGRAM
108.177.127.113 RAW
108.177.127.102 STREAM
108.177.127.102 DGRAM
108.177.127.102 RAW
108.177.127.101 STREAM
108.177.127.101 DGRAM
108.177.127.101 RAW
108.177.127.138 STREAM
108.177.127.138 DGRAM
108.177.127.138 RAW

$ docker run -it registry.k8s.io/ubuntu-slim:0.14 getent hosts google.com
2a00:1450:4013:c14::64 google.com
2a00:1450:4013:c14::71 google.com
2a00:1450:4013:c14::65 google.com
2a00:1450:4013:c14::66 google.com
```

Reply:
Guys, it's very old images

```
registry.k8s.io/busybox                                     1.27.2                             54511612f1c4   7 years ago    1.13MB
registry.k8s.io/ubuntu-slim                                 0.14                               fba1281b32ff   7 years ago    43.6MB
registry.k8s.io/alpine-with-bash                            1.0                                822c13824dc2   9 years ago    6.67MB
```

I still prefer to create separate image.

Reply:
I agree kjobctl should have its own image for this, called something like kjobctl-slurm-init. 

However there is some overhead on publishing the image and decisions on how it should be released. And the release process might change once kjobctl is separated from Kueue .

So, if we are close to making it work with busybox, then let's complete and we will open a new issue to use our image as a follow up. This way we don't block the functionality on this, and can prioritize it independently.

Reply:
SGTM 🖖

Reply:
Ok, already opened actually: https://github.com/kubernetes-sigs/kueue/issues/3243.

Reply:
OK. Looks like working fine with busybox solution.